# Recursive Function

Recursive functions are functions that call themselves during their execution. They are useful for solving problems that can be broken down into smaller, similar subproblems. Here's a step-by-step explanation of how recursive functions work in Python.

- **Base Case:** Every recursive function needs a base case, which is the simplest possible case where the function does not call itself. It acts as the termination condition for the recursion. Without a base case, the recursive function would continue calling itself indefinitely, resulting in an infinite loop.

- **Recursive Case:** In addition to the base case, recursive functions also have a recursive case where the function calls itself with a modified version of the problem. This recursive call allows the function to break down a larger problem into smaller subproblems until it reaches the base case.

Now, let's look at an example to illustrate the concept of recursive functions in Python. We'll implement a factorial function using recursion.

In [1]:
def factorial(n):
    if n == 0:  # Base case: factorial of 0 is 1
        return 1
    else:
        return n * factorial(n - 1)  # Recursive case: n! = n * (n-1)!

print(factorial(5))

120


In this example, the `factorial` function calculates the factorial of a given number `n`. If `n` is equal to `0`, it returns `1`, which is the base case. Otherwise, it calls itself with `n-1` and multiplies the result by `n`, which is the recursive case.

The function starts with `factorial(5)`. Since `5` is not `0`, it goes to the recursive case and calls `factorial(4)`. Again, `4` is not `0`, so it calls `factorial(3)`. This process continues until `factorial(0)` is reached, which returns `1`. The intermediate results are then multiplied together, resulting in `factorial(5) = 5 * 4 * 3 * 2 * 1 = 120`.

Recursive functions can be powerful, but it's important to ensure they have a base case and that the recursive calls converge towards the base case. Otherwise, the function may run indefinitely and cause a stack overflow error.